src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+        // as the commit offset won't change when partitions are produced to
+        // without acks=all. The `last_visible_index` more closely corresponds
+        // to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes{ctx.requests.size()};
+
+        for (size_t i = 0; i < ctx.requests.size(); i++) {
+            const auto& req = ctx.requests[i];
+            auto part = ctx.mgr.get(req.ktp());
+            if (!part) {
+                errored_partitions.push_back(i);
+                continue;
+            }
+            auto consensus = part->raft();
+            if (!consensus) {
+                errored_partitions.push_back(i);
+                continue;
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
+        }
+
+        // Fast-path fetch
+        std::vector<read_result> results = co_await fetch_ntps_in_parallel(
+          ctx.mgr,
+          ctx.srv.get_replica_selector(),
+          std::move(requests),
+          ctx.foreign_read,
+          ctx.deadline,
+          ctx.bytes_left,
+          ctx.srv.memory(),
+          ctx.srv.memory_fetch_sem());
+
+        // A read_result needs to be returned for every partition. Hence, the
+        // function can't return before the first call to
+        // `fetch_ntps_in_parallel`. However, if we weren't able to read the
+        // last_visible_index for a partition before the first call to
+        // `fetch_ntps_in_parallel` then we need to return with an error for
+        // that partition.
+        for (auto i : errored_partitions) {
+            results[i] = read_result(error_code::not_leader_for_partition);
+            results[i].partition = ctx.requests[i].ktp().get_partition();
+        }
+        errored_partitions.clear();
+
+        for (size_t i = 0; i < results.size(); i++) {
+            const auto& r = results[i];
+            total_size += r.data_size_bytes();
+            if (r.error != error_code::none) {
+                errored_partitions.push_back(i);
+            }
+        }
+
+        if (
+          total_size > ctx.min_bytes || !errored_partitions.empty()
+          || ctx.as.abort_requested()) {
+            co_return worker_result{
+              .read_results = std::move(results),
+              .errored_partitions = std::move(errored_partitions),
+              .total_size = total_size,
+            };
+        }
+
+        // If the fast-path doesn't return there isn't enough data so
+        // we'll have to wait for the last_visible_index of partitions to
+        // change.
+        ss::gate waiter_gate;
+        ss::condition_variable has_requests;
+        std::vector<size_t> request_indices;
+
+        // An abort source is needed to clear any waiters from the
+        // `raft::offset_monitor` after the worker returns.
+        ss::abort_source waiter_as;
+        auto hwm_as_sub = _as.subscribe([&waiter_as, &has_requests]() noexcept {
```

Comment:
AFAICT this an `optimized_optional<subscripton>` and if it is returned empty the as was already triggered _and the callback won't occur_, so in this case you'd end up doing the rest of the work and waiting even though you should return. So I think you need to check the value of subscribe here and elsewhere and early out if unengaged.

Reply:
Nice catch, now checking the optional in the equivalent subscription in `fetch_woker::run`